Building the simple recommender is fairly straightforward. The
steps are as follows:
1. Choose a metric (or score) to rate the movies on
2. Decide on the prerequisites for the movie to be featured
on the chart
3. Calculate the score for every movie that satisfies the
conditions
4. Output the list of movies in decreasing order of their
scores

In [1]:
import pandas as pd

In [2]:
movies = pd.read_csv('tmdb_5000_movies.csv')
credits = pd.read_csv('tmdb_5000_credits.csv')

In [3]:
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [7]:
movies.shape

(4803, 20)

Weighted Rating (WR) = ((v/(v+m))*R)+((v/(v+m))*C)

v is the number of votes garnered by the movie
m is the minimum number of votes required for the
movie to be in the chart (the prerequisite)
R is the mean rating of the movie
C is the mean rating of all the movies in the dataset

In [5]:
m = movies['vote_count'].quantile(0.80)
m

957.6000000000004

We can see that only 20% of the movies have gained more than 957.6000000000004 votes. Therefore, our value of m is 957.6000000000004.

Another prerequisite that we want in place is the runtime. We
will only consider movies that are greater than 45 minutes and
less than 300 minutes in length. Let us define a new DataFrame,
q_movies, which will hold all the movies that qualify to appear in
the chart

In [6]:
q_movies = movies[(movies['runtime'] >= 45) & (movies['runtime'] <= 300)]
q_movies.shape

(4761, 20)

We see that from our dataset of 4803 movies approximately 4761 movies made the cut.

The final value that we need to discover before we calculate our
scores is C, the mean rating for all the movies in the dataset

In [8]:
C = movies['vote_average'].mean()
C

6.092171559442016

We can see that the average rating of a movie is approximately
6.09/10. It seems that IMDB happens to be particularly strict
with their ratings. Now that we have the value of C, we can go
about calculating our score for each movie

First, let us define a function that computes the rating for a
movie, given its features and the values of m and C

In [9]:
def weighted_rating(x, m=m, C=C):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [10]:
q_movies['score'] = q_movies.apply(weighted_rating, axis=1)

C:\Users\lenovo\AppData\Local\Temp\ipykernel_8032\3735737523.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  q_movies['score'] = q_movies.apply(weighted_rating, axis=1)


There is just one step left. We now need to sort our DataFrame
on the basis of the score we just computed and output the list of
top movies

In [12]:
q_movies.sort_values(by='score', ascending=False)

budget                                             genres  \
1881   25000000  [{"id": 18, "name": "Drama"}, {"id": 80, "name...   
662    63000000                      [{"id": 18, "name": "Drama"}]   
3337    6000000  [{"id": 18, "name": "Drama"}, {"id": 80, "name...   
3232    8000000  [{"id": 53, "name": "Thriller"}, {"id": 80, "n...   
65    185000000  [{"id": 18, "name": "Drama"}, {"id": 28, "name...   
...         ...                                                ...   
303   100000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
3746    4000000                   [{"id": 53, "name": "Thriller"}]   
1652  100000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
210   125000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
242   120000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                          homepage      id  \
1881                                           NaN     278   
662     http://www.foxmovies.com/movies/fight-club     550   
3337                  http://www.thegodfather.com/     238   
3232                                           NaN     680   
65    http://thedarkknight.warnerbros.com/dvdsite/     155   
...                                            ...     ...   
303                                            NaN     314   
3746            http://www.theboynextdoorfilm.com/  241251   
1652                                           NaN   14164   
210                                            NaN     415   
242             http://www.fantasticfourmovie.com/  166424   

                                               keywords original_language  \
1881  [{"id": 378, "name": "prison"}, {"id": 417, "n...                en   
662   [{"id": 825, "name": "support group"}, {"id": ...                en   
3337  [{"id": 131, "name": "italy"}, {"id": 699, "na...                en   
3232  [{"id": 396, "name": "transporter"}, {"id": 14...                en   
65    [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
...                                                 ...               ...   
303   [{"id": 418, "name": "white russian"}, {"id": ...                en   
3746  [{"id": 255, "name": "male nudity"}, {"id": 29...                en   
1652  [{"id": 3436, "name": "karate"}, {"id": 9715, ...                en   
210   [{"id": 848, "name": "double life"}, {"id": 84...                en   
242   [{"id": 3822, "name": "teleportation"}, {"id":...                en   

                original_title  \
1881  The Shawshank Redemption   
662                 Fight Club   
3337             The Godfather   
3232              Pulp Fiction   
65             The Dark Knight   
...                        ...   
303                   Catwoman   
3746         The Boy Next Door   
1652      Dragonball Evolution   
210             Batman & Robin   
242             Fantastic Four   

                                               overview  popularity  \
1881  Framed in the 1940s for the double murder of h...  136.747729   
662   A ticking-time-bomb insomniac and a slippery s...  146.757391   
3337  Spanning the years 1945 to 1955, a chronicle o...  143.659698   
3232  A burger-loving hit man, his philosophical par...  121.463076   
65    Batman raises the stakes in his war on crime. ...  187.322927   
...                                                 ...         ...   
303   Liquidated after discovering a corporate consp...   32.271938   
3746  A recently cheated on married woman falls for ...   24.161735   
1652  The young warrior Son Goku sets out on a quest...   21.677732   
210   Along with crime-fighting partner Robin and ne...   50.073575   
242   Four young outsiders teleport to a dangerous u...   38.126095   

                                   production_companies  ... release_date  \
1881  [{"name": "Castle Rock Entertainment", "id": 97}]  ...   1994-09-23   
662   [{"name": "Regency Enterprises", "id": 508}, {...  ...   1999-10-15   
3337 